[PC-1809] Alvik - User Manual Docs (#2064)
arduino/docs-content: content/hardware/08.edu/solution-and-kits/alvik/tutorials/user-manual/user-manual.md

description: "Discover how to use and program Arduino® Alvik's robot"
tags:
  - Robot

> author: "Paolo Cavagnolo, Pedro Miguel"

Reviewer: ```suggestion
author: "Paolo Cavagnolo, Pedro Miguel, Jose Garcia"

```

---

## Introduction
Arduino® Alvik is a powerful and versatile robot specifically designed for programming and STEAM education.

![Alvik's Robot](assets/alvik_main.jpg)

Powered by the Arduino® Nano ESP32, Alvik offers diverse learning paths through different programming languages including MicroPython, Arduino C, and block-based coding; enabling different possibilities to explore Robotics, IoT, and Artificial Intelligence.

In this tutorial, you will find useful information to get started, test, and maintain Alvik.

> ## What Is Alvik

Reviewer: ```suggestion
## What Is Alvik

```

Reviewer: The title is nor rendering properly

Reviewer 2: Will check if this is still the case

Alvik is a robot with two controllers and a wide range of sensors and actuators. The main controller is the Arduino Nano ESP32 attached to the top of the robot while there is an STM32 controller integrated into the robot that takes care of the low-level commands such as reading the sensors and moving the motors. It has been designed for STEAM education, making it an ideal tool for learning programming, Robotics, IoT, Artificial Intelligence...

### Unboxing

> ![Unboxing Alvik](assets/unboxing.jpg)

Reviewer: missing image

Inside the Arduino Alvik package you'll find the following items:

- Alvik robot
- 18650 Li-Ion battery attached to the robot
- USB-C® to USB-C® cable

### Alvik Overview

![Alvik exploded view](assets/main-components.png)

***When the PCB is out of the chassis and the battery is in place there is the risk of short-circuiting the 18650 Li-Ion battery. If you remove the hardware from the chassis make sure you do it on a __non-conductive surface clean__ of materials or tools that can short-circuit the battery***

### Main Components

![Alvik top components](assets/up-components.png)

![Alvik bottom components](assets/down-components.png)

> ## The Brain

Reviewer: ```suggestion
## Alvik's Brain

```

![Alvik Brain](assets/alvik_brain.png)

### Nano ESP32

The [Nano ESP32](https://store.arduino.cc/products/nano-esp32) is the board used to control Alvik. It has a fast processor, large flash memory, and a Wi-Fi® enabled chip packed into a tiny circuit board.

***You can find out more about this board in the [Nano ESP32 documentation](https://docs.arduino.cc/hardware/nano-esp32).***

Please note that when using MicroPython the pin number reflects the GPIO on the ESP32-S3, not the Nano board. Use the **green labeled number** in the following image. You can read more about this [here](https://docs.arduino.cc/micropython/micropython-course/course/introduction-python#nano-esp32--micropython-pinout).

![Nano ESP32 pinout](assets/esp-pinout.png)

> ## The Body

Reviewer: ```suggestion
## Alvik's Body

```

***You can find out more about this board in the [Nano ESP32 documentation](https://docs.arduino.cc/hardware/nano-esp32).***

Please note that when using MicroPython the pin number reflects the GPIO on the ESP32-S3, not the Nano board. Use the **green labeled number** in the following image. You can read more about this [here](https://docs.arduino.cc/micropython/micropython-course/course/introduction-python#nano-esp32--micropython-pinout).

![Nano ESP32 pinout](assets/esp-pinout.png)

## The Body

> Using the same analogy, the body of the Alvik robot includes everything else from the sensors and motors to the connectors supporting external modules and, of course, the STM32, which acts like the nervous system of the robot. Just as the human nervous system transmits signals between the brain and various parts of the body, the STM32 transmits commands and collects data from the robot's components. When the Nano ESP32, acting as the brain, sends commands, the STM32 executes these low-level operations, such as moving the motors, flashing the LEDs, or interacting with external devices. The STM32 is essential for carrying out the detailed actions based on the high-level instructions from the Nano ESP32, ensuring that the robot's movements and functions are performed accurately.

Reviewer 2: Created a task on this:
https://arduino.atlassian.net/browse/MKC-1755

### STM32

> The main core of the robot is the STM32 ARM Cortex-M4 32 Bit controller. It handles the low-level commands such as reading sensors and moving motors. You can access it through a set of dedicated APIs from the Nano ESP32.

Reviewer 2: Added image

> ***Although Alvik's "out of the box" experience is designed in micropython, there are C++ APIs that allow you to program Avlik in the Arduino IDE, these APIs are 1 to 1 compatible with the micropython APIs, which means that the API overview will be interesting for both micropython and C++ programs.***

Reviewer: ```suggestion
***Although Alvik's "out of the box" experience is designed in micropython, there are C++ APIs that allow you to program Avlik in the Arduino IDE, these APIs are 1 to 1 compatible with the micropython APIs, which means that the [Alvik's API Documentation](https://docs.arduino.cc/tutorials/alvik/api-overview) will give you information for both micropython and C++ programming experiences.***

```

![Alvik Body](assets/alvik_body.png)

### Controller - STM32

The main core of the robot is the STM32 ARM Cortex-M4 32 Bit controller. It handles the low-level commands such as reading sensors and moving motors. You can access it through a set of dedicated APIs from the Nano ESP32.

You can learn more about the available functions for Alvik in the following [Alvik's API Documentation](https://docs.arduino.cc/tutorials/alvik/api-overview).

***Although Alvik's "out of the box" experience is designed in micropython, there are C++ APIs that allow you to program Avlik in the Arduino IDE, these APIs are 1 to 1 compatible with the micropython APIs, which means that the [Alvik's API Documentation](https://docs.arduino.cc/tutorials/alvik/api-overview) will give you information for both micropython and C++ programming experiences.***

> The latest firmware of the STM32 can be found at [this link](https://github.com/arduino-libraries/Arduino_AlvikCarrier/releases), and [here](#how-to-upload-firmware) is the guide to flash it.

Reviewer 2: checked page is working outside preview

> ![Notification to switch on](assets/message-switch-on.png)

Reviewer 2: Added blur and a colored highlight. I added in orange to avoid giving the idea that it is a error but let me know if you prefer red.

### Battery

> The battery is a rechargeable Li-ion 18650. It is located in the bottom part of Alvik, to access it you need to remove one Phillip's screw and take out the plastic holder.

Reviewer: we need to add here information about how to proceed when replacing the battery. Users will need to:
- Detach existing battery
- Attaching a new battery
- Plug NanoESP32 (in case they are using other controller or using the robot in a different manner)
- Plug the NanoESP32 to the computer
Unless users follow these steps, the robot won't work after a battery replacement, these steps are needed to reset the BMS after a battery change. If users don't do this, the BMS is "off" and the robot won't turn on

Reviewer 2: Added

***Don't confuse the RGB status LED with the power ON LED of the Nano ESP32, which is always green.***

> ## Sensors

Reviewer: ```suggestion
## Alvik's Sensors

```

***Before using the ToF sensor check if it has a yellow protective film, if present, remove it from the sensor to ensure it works properly***

### Touch Buttons

The Arduino Alvik robot is equipped with seven capacitive touch buttons. These buttons allow for interactive input and can be used to control various functions of the robot, such as movement, mode selection, and more.

**Available buttons are:**
- Up
- Down
- Left
- Right
- Center
- OK (represented by a "tick" Icon)
- Cancel (represented by a "x")

![Touch Button position](assets/touchButton.png)

Each button can be programmed to perform specific actions, providing more complex interface for user interaction.

### Distance Sensors

The Arduino Alvik robot is equipped with a ToF (Time of Flight) 8x8 Array sensor, specifically the VL53L7CX, which can measure distances up to 350 cm. These distance sensors help the robot detect obstacles and measure the distance to objects in its environment. The sensor is integrated into the front PCB of the robot.

**Key Features:**
- Maximum distance: 350 cm
- High precision and accuracy
- Used for obstacle detection and distance measurement

> ![Distance sensor position](assets/distanceSensor.png)

Reviewer 2: for the section ### Distance Sensor I will try to get a visual as all we have are 2D ones

- Measures acceleration, angular velocity, and orientation
- Provides roll, pitch, and yaw values
- Useful for motion tracking and stabilization

![IMU axis reference](assets/imageIMU.png)

## Actuators

### Motors and Encoders

Alvik has two high-precision geared motors and two RGB LEDs. The motors are driven by the MAX22211 motor driver and include magnetic relative encoders for precise control.

| **Actuator name**        | **Part name**           | **Test program name** |
|--------------------------|-------------------------|-----------------------|
| Geared motors w/ encoder | GM12-N20VA-08255-150-EN | wheels_positions.py   |
| RGB LEDs                 | RGB LEDs                | leds_settings.py      |

The RGB LEDs can be used for visual feedback and can be programmed to display various colors and patterns.

### Connectors

The connectors are placed in the back of the robot, the pinout is shown in the following image:

![Connectors Pinout](assets/datasheet_connectors.png)

## What the Robot Includes

The Arduino Alvik robot package includes the following items:
- Alvik robot
- USB-C to USB-C cable
- Phillips screwdriver
- 18650 Li-Ion battery

## Move Alvik!

> ### Basic Touch Programming

Reviewer: This comment still valid IMO we are not adding any valuable information about the three pre-flashed activities, if we want to keep this section either I suggest to explain the logic about how the OoB experience is built:
- 3 Different programs saved in different ".py" files
- 1 demo program handling the buttons behavior
- 1 main.py calling the demo

If we want to offer info about the programs, I think we should link to the GS guide, IMO is more complete.

I don't understand what the user is getting from this section as it is.

> ### Basic Touch Programming

Reviewer 2: Created a small list format and linked to the GS guide for more info.
This way people get the idea of the included sketch but we dont go over unecessary info

***Before using the ToF sensor check if it has a yellow protective film, if present, remove it from the sensor to ensure it works properly***

### Touch Buttons

The Arduino Alvik robot is equipped with seven capacitive touch buttons. These buttons allow for interactive input and can be used to control various functions of the robot, such as movement, mode selection, and more.

**Available buttons are:**
- Up
- Down
- Left
- Right
- Center
- OK (represented by a "tick" Icon)
- Cancel (represented by a "x")

![Touch Button position](assets/touchButton.png)

Each button can be programmed to perform specific actions, providing more complex interface for user interaction.

> ### Distance Sensors

Reviewer: ```suggestion
### Distance Sensor

```

Reviewer 2: Good point

The Arduino Alvik robot is equipped with a ToF (Time of Flight) 8x8 Array sensor, specifically the VL53L7CX, which can measure distances up to 350 cm. These distance sensors help the robot detect obstacles and measure the distance to objects in its environment. The sensor is integrated into the front PCB of the robot.

**Key Features:**
- Maximum distance: 350 cm
- High precision and accuracy
- Used for obstacle detection and distance measurement

![Distance sensor position](assets/distanceSensor.png)

> ### Line Follower Sensor

Reviewer: ```suggestion
### Line Follower Sensors

```

### Color Sensor

The color sensor on the Arduino Alvik robot is used to detect and identify colors on surfaces that the robot encounters. It provides both raw color readouts and labeled color information that can be used for various applications such as line following, object detection, and more. The sensor used is the APDS 9660.

**Key Features:**
- Detects and identifies colors
- Provides raw color readouts and labeled color information
- Useful for line following, object detection, and other applications

![Color sensor bottom view](assets/imageColor.png)

### IMU

The Arduino Alvik robot is equipped with an onboard IMU (Inertial Measurement Unit) that provides valuable information about the robot's motion and orientation. The IMU can measure acceleration, angular velocity, and orientation. The sensor used is the LSM6DSOX.

**Key Features:**
- Measures acceleration, angular velocity, and orientation
- Provides roll, pitch, and yaw values
- Useful for motion tracking and stabilization

![IMU axis reference](assets/imageIMU.png)

> ## Actuators

Reviewer: ```suggestion
## Alvik's Actuators

```

| **Actuator name**        | **Part name**           | **Test program name** |
|--------------------------|-------------------------|-----------------------|

> | Geared motors w/ encoder | GM12-N20VA-08255-150-EN | wheels_positions.py   |

Reviewer: It feel weird we don't have information about the motors, we have info about each component independently but not about the motors, one of the key components of the robot.
We should add info about them, e.g. speed, torque, consumption(?), voltage(?), gear ratio...
and also for the encoders, e.g. type of encoder, pulses per turn, etc.

Reviewer: missing also visual pointing/showing the LEDs, we have visuals for all the components individually...
as well as title for each one of the "actuators" "motors" and "LEDs"

Reviewer 2: Image created

Reviewer 2: Added some metrics for the motors

| RGB LEDs                 | RGB LEDs                | leds_settings.py      |

> The RGB LEDs can be used for visual feedback and can be programmed to display various colors and patterns. The LEDs take in **boolean** values for each color meaning that combinations are possible by setting each color **on** or **off** insteadof controlling intensity.

Reviewer: ```suggestion
The RGB LEDs can be used for visual feedback and can be programmed to display various colors and patterns. The LEDs take in **boolean** values for each color meaning that combinations are possible by setting each color **on** or **off** .

```

### Connectors

The connectors are placed in the back of the robot, the pinout is shown in the following image:

![Connectors Pinout](assets/datasheet_connectors.png)

> ## What the Robot Includes

Reviewer: all this section is at the beginning of the tutorial, under the unboxing, I don't see the point of repeating this here

> ## What the Robot Includes

Reviewer: I would rename this section something else, we are talking here about Out of the box experience, my perception about "what the robot includes" is more related to physical aspects

### Following Objects

Out of the box, Alvik is equipped with distance sensors that allow it to follow objects by maintaining a certain distance from them. This feature is pre-configured to help the Alvik follow a target object, making it useful for applications like automated guided vehicles and interactive robotics. The default settings ensure that Alvik can smoothly track and follow objects placed in front of it without any additional programming required.

### Line Follower

Alvik comes pre-programmed with a line following algorithm that utilizes its built-in line follower sensors. These sensors enable the Alvik to navigate along predefined paths marked by lines on the ground. This feature is particularly beneficial in educational settings and for demonstrating basic robotic navigation. The default configuration allows Alvik to follow a black line on a white surface right out of the box, providing an engaging and interactive learning experience for you.

## Alvik's Firmware

To get started to play with Alvik you will need the following hardware and software:

> ### Hardware Requirements

Reviewer: this section is repeated 3 times in the manual, I think having it on the 1st section, in the unboxing, is enough

Reviewer 2: Removed multiple listing of the components this is indeed a good point!

- Alvik (x1)
- USB-C® to USB-C® cable (x1)
- Phillips Screwdriver (cross head)(optional) (x1)
- Computer (x1)

***Make sure the USB-C® cable you are using works with data lines, not only power lines***

### Software Requirements

- Operating Systems: Linux, macOS or Windows

- [Arduino Lab for Micropython](https://labs.arduino.cc/en/labs/micropython)

### Updating Alvik's Brain (Nano ESP32)

It may happen that you used your Nano ESP32 for other projects than Alvik, or you need to replace it. In order to make your Nano ESP32 work with Alvik, there are few steps needed:

1. Install the micropython bootloader on it following [this guide](https://docs.arduino.cc/micropython/basics/board-installation/).

2. Download the Alvik micropyton libraries

>     Alvik micropython libraries from the [Alvik repository](https://github.com/arduino/arduino-alvik-mpy/tree/main)

Reviewer: ```suggestion
    - Alvik micropython libraries from the [Alvik repository](https://github.com/arduino/arduino-alvik-mpy/tree/main)

```

>     ucPack libraries from the [ucPack repository](https://github.com/arduino/ucPack-mpy/tree/main)

Reviewer: ```suggestion
   -  ucPack libraries from the [ucPack repository](https://github.com/arduino/ucPack-mpy/tree/main)

```

        turning()
    else:
        alvik.set_wheels_speed(speed, speed)

```

**4. **Connect Alvik to your PC using the cable included in the box, under the tray.

![Connecting Alvik to the PC](assets/connecting_alvik.gif)

***Make sure that Alvik is OFF before connecting it to your computer.***

**5. **Once Alvik is connected to the PC, connect it to the Arduino Lab for MicroPython and open the _main.py_ file in the Alvik folder. Once the file is opened let's replace the `import demo` statement by `import obstacle_avoider`.

![Connecting Alvik to the IDE](assets/connecting_alvik_ide.gif)

***If you want to go back to the out of the box experience where you could select between reg, green and blue programs, you only need to modify the _main.py_ again replacing the `import obstacle_avoider` statement by `import demo`***

**6. **The last step is to move the _obstacle_avoider.py_ file from the local repository to Alvik's memory.

![Moving file from local to Alvik's memory](assets/local2memory.gif)

You are now all set, disconnect Alvik from the computer, put some obstacles around Alvik, turn it ON and see how Alvik detects them and turns to avoid them.

> ## Deep Dive Programming Alvik

Reviewer: ```suggestion
## Deep Dive Programming Alvik

```

   The `set_color` function sets the color of the LED by specifying the intensity of red, green, and blue components.

   **Inputs:**
    `red`: Intensity of the red component (0 or 1)
    `green`: Intensity of the green component (0 or 1)
    `blue`: Intensity of the blue component (0 or 1)

In this case the LEDs take in **boolean** values for each color meaning that combinations are possible by setting each color **on** or **off** insteadof controlling intensity.

**Example**

Here's an example sketch that cycles through different colors on both LEDs:

```python
from arduino_alvik import ArduinoAlvik
from time import sleep
import sys

alvik = ArduinoAlvik()
alvik.begin()
sleep(5)

while True:
  alvik.left_led.set_color(0, 0, 0)
  alvik.right_led.set_color(0, 0, 0)
  sleep(1)
  alvik.left_led.set_color(0, 0, 1)
  alvik.right_led.set_color(0, 0, 1)
  sleep(1)
  alvik.left_led.set_color(0, 1, 0)
  alvik.right_led.set_color(0, 1, 0)
  sleep(1)
  alvik.left_led.set_color(1, 0, 0)
  alvik.right_led.set_color(1, 0, 0)
  sleep(1)
```

By following these steps, you can control the LEDs on the Arduino Alvik robot to display different colors and create various lighting effects.

> ## Talking With Other Machines!

Reviewer: ```suggestion
## Talking With Other Devices!

```

### WiFi

The ESP32 on the Arduino Alvik robot includes built-in Wi-Fi capabilities, enabling it to connect to wireless networks and communicate with other devices over the internet. This can be particularly useful for remote control applications. In this example, we'll set up the Alvik robot to connect to a Wi-Fi network, host a web server, and provide a simple web interface with buttons to control the robot's movements.

The provided code will:
1. Connect the ESP32 to a specified Wi-Fi network.
2. Set up a web server to listen for incoming HTTP requests.
3. Serve a simple web page with buttons for directional control.
4. Move the robot in the specified direction when a button is pressed.

> ### Instructions

Reviewer: ```suggestion
### Step by Step setup

```

**Device Identification**

It is important for ESP-NOW to ensure that youknow the MAC address for your device.

You can do that by running the following code:
```python
import network

# Initialize the network interface
e = network.WLAN(network.STA_IF)
e.active(True)

# Get and print the MAC address
mac = e.config('mac')
print("MAC address:", ':'.join('%02x' % b for b in mac))
```
Each device will have a different address, feel free to identify your devices so that you can easily build your ESP-NOW projects.

Now that you know the address for your device we can run your first ESP-NOW project.

Lets now establish connection between a pair of Alviks. One will act as a controller and the other will follow the controls and move in the according direction.

The provided code will:
1. Initialize ESP-NOW on the ESP32.
2. Set up a callback function to handle incoming ESP-NOW messages.
3. Move the robot based on the received commands while buttons are pressed.

> ### Instructions

Reviewer: ```suggestion
### Step by Step setup

```

1. Upload the receiver script to the Arduino Nano ESP32 on the Alvik robot.
2. Use another ESP32 device to send commands to the Alvik robot via ESP-NOW.

> ### Code

Reviewer: ```suggestion

```

import espnow

# Initialize ESP-NOW
wlan = network.WLAN(network.STA_IF)
wlan.active(True)
esp = espnow.ESPNow()
esp.active(True)

# Print the MAC address
mac = wlan.config('mac')
print("MAC address:", ':'.join('%02x' % b for b in mac))

# Initialize Alvik
alvik = ArduinoAlvik()
alvik.begin()

def perform_action(command):
    print(f"Performing action: {command}")
    if command == "MOVE_FORWARD":
        alvik.set_wheels_speed(60, 60)
    elif command == "MOVE_BACKWARD":
        alvik.set_wheels_speed(-60, -60)
    elif command == "TURN_LEFT":
        alvik.set_wheels_speed(-60, 60)
    elif command == "TURN_RIGHT":
        alvik.set_wheels_speed(60, -60)
    elif command == "STOP":
        alvik.brake()
    elif command == "ACTION":
        pass

print("Waiting for commands...")

while True:
    try:
        host, msg = esp.recv()
        if msg:
            print(f"Received message from {host}: {msg}")
            perform_action(msg.decode())
        else:
            print("No message received")
    except Exception as e:
        print(f"An error occurred: {e}")
    except KeyboardInterrupt:
        alvik.stop()
        print("Script terminated by user")
        sys.exit()
```

> ### Sending Commands

Reviewer: ```suggestion
**Setting Up ESP-NOW on the Transmitter**
```

button_commands = {
    "UP": "MOVE_FORWARD",
    "DOWN": "MOVE_BACKWARD",
    "LEFT": "TURN_LEFT",
    "RIGHT": "TURN_RIGHT",
    "OK": "STOP",
    "CANCEL": "STOP",
    "CENTER": "ACTION"
}

def send_command(command):
    esp.send(peer_mac, command.encode())
    print(f"Sent command: {command}")

print("Controller running. Press buttons to send commands.")

while True:
    try:
        if alvik.get_touch_any():
            if alvik.get_touch_up():
                send_command(button_commands["UP"])
            if alvik.get_touch_down():
                send_command(button_commands["DOWN"])
            if alvik.get_touch_left():
                send_command(button_commands["LEFT"])
            if alvik.get_touch_right():
                send_command(button_commands["RIGHT"])
            if alvik.get_touch_ok():
                send_command(button_commands["OK"])
            if alvik.get_touch_cancel():
                send_command(button_commands["CANCEL"])
            if alvik.get_touch_center():
                send_command(button_commands["CENTER"])

        sleep_ms(100)
    except KeyboardInterrupt:
        alvik.stop()
        sys.exit()
```

> ### Receiver

Reviewer: ```suggestion
**Receiver behaviour**

```

1. **Initializing ESP-NOW**: The code sets up the ESP32 as a Wi-Fi station and initializes ESP-NOW.
2. **Callback Function**: A callback function `on_receive_msg` is defined to handle incoming messages. This function decodes the message and moves the robot based on the received command.
3. **Registering the Callback**: The callback function is registered with ESP-NOW to ensure it gets called when a message is received.

> ### Sender

Reviewer: ```suggestion
**Transmitter behaviour**

```

1. **Initializing ESP-NOW**: The code sets up the ESP32 as a Wi-Fi station and initializes ESP-NOW.
2. **Adding a Peer**: The MAC address of the receiver (Alvik robot) is added as a peer.
3. **Sending Commands**: The function `send_command` sends commands to the receiver. Example commands are provided to move the robot in different directions.

By following these steps, you'll be able to control your Arduino Alvik robot remotely using ESP-NOW, a fast and efficient communication protocol.
Feel free to tinker with this feature to add new and interesting interactions.

> ## Maintenance

Reviewer: Couple of things in this section:
1. Please let's get rid of the "Check wheel alignment" image XD and replace it for another one, render, IDK something better looking 
2. 1st snippet on the "Calibrate Color Sensors" is incomplete or has bugs

Reviewer 2: This is from the original document I will have to request a visual for this and re-test

## Out Of The Box Modes

Alvik is equipped with three ready-to-go examples. To choose one of the examples, just turn your Alvik ON, move the switch located at the bottom right corner of the robot to the right, wait until the LEDs turn blue and use the Up and Down buttons to pick one color, then hit the "tick" confirmation button. 

- **Basic Touch Programming**: Alvik’s touch buttons can control actions such as movement, changing modes, and interacting with the environment. This allows for intuitive control schemes for different applications.

- **Following Objects**: Using its distance sensors, Alvik can follow objects by maintaining a specific distance. This feature is useful for automated guided vehicles and interactive robotics.

- **Line Follower**: Alvik's line follower sensors enable it to navigate along paths marked by lines on the ground, which is great for educational purposes and demonstrating basic robotic navigation.

For more information, visit the [Alvik Getting Started Guide](https://docs.arduino.cc/tutorials/alvik/getting-started/).

## Alvik's Firmware

> To get started to play with Alvik you will need the following hardware and software:

Reviewer: ```suggestion
To get started to play with Alvik you will need the following software:

```

### Software Requirements

- Operating Systems: Linux, macOS or Windows

- [Arduino Lab for Micropython](https://labs.arduino.cc/en/labs/micropython)

### Updating Alvik's Brain (Nano ESP32)

It may happen that you used your Nano ESP32 for other projects than Alvik, or you need to replace it. In order to make your Nano ESP32 work with Alvik, there are few steps needed:

1. Install the micropython bootloader on it following [this guide](https://docs.arduino.cc/micropython/basics/board-installation/).

2. Download the Alvik micropyton libraries

>     - Alvik micropython libraries from the [Alvik repository](https://github.com/arduino/arduino-alvik-mpy/tree/main)

Reviewer: This is not rendered properly in the preview

>    -  ucPack libraries from the [ucPack repository](https://github.com/arduino/ucPack-mpy/tree/main)

Reviewer: This is not rendered properly in the preview

> 3. Unzip both of the downloaded libraries in a single "Alvik" folder, open the Arduino Lab for MicroPython, go to the "files" tab and set the path to the unzipped folder on the Arduino Lab for Micropython

Reviewer: this should be #3 but is rendering as #1 again

**Inputs:**
- `linear_unit`: Unit of linear velocity (default is 'cm/s').
- `angular_unit`: Unit of rotational speed of the wheels (default is 'deg/s').

**Outputs:**
- `linear_velocity`: The linear speed of the robot.
- `angular_velocity`: The angular speed of the wheels.

**Example Usage:**
```python
linear_velocity, angular_velocity = alvik.get_drive_speed(linear_unit='cm/s', angular_unit='deg/s')
print(f"Linear Velocity: {linear_velocity} cm/s, Angular Velocity: {angular_velocity} deg/s")
```

**Check Pose**

The `get_pose` function returns the current pose of the robot.

**Inputs:**
- `distance_unit`: Unit of x and y outputs (default is 'cm').
- `angle_unit`: Unit of theta output (default is 'deg').

**Outputs:**
- `x`: The x-coordinate of the robot.
- `y`: The y-coordinate of the robot.
- `theta`: The orientation angle of the robot.

**Example Usage:**
```python
x, y, theta = alvik.get_pose(distance_unit='cm', angle_unit='deg')
print(f"Pose - X: {x} cm, Y: {y} cm, Theta: {theta} degrees")
```

These functions allow for precise control and monitoring of the Alvik robot's movements, enabling advanced navigation and path planning capabilities.

> ### Reading Buttons

Reviewer: the links in this section are not working

Reviewer 2: The links for the buttons are working when not in preview

**Example**
The following example demonstrates how to read the state of each button and print its name when pressed:

```python
from arduino_alvik import ArduinoAlvik
from time import sleep_ms
import sys

alvik = ArduinoAlvik()
alvik.begin()

while True:
    try:
        if alvik.get_touch_any():
            if alvik.get_touch_up():
                print("UP")
            if alvik.get_touch_down():
                print("DOWN")
            if alvik.get_touch_left():
                print("LEFT")
            if alvik.get_touch_right():
                print("RIGHT")
            if alvik.get_touch_ok():
                print("OK")
            if alvik.get_touch_cancel():
                print("CANCEL")
            if alvik.get_touch_center():
                print("CENTER")

        sleep_ms(100)
    except KeyboardInterrupt as e:
        print('over')
        alvik.stop()
        sys.exit()
```

This example provides a simple way to read and respond to button presses on the Arduino Alvik robot.
Feel free to expand to it by adding customized responses to each different press.

> ### Detecting Obstacles

Reviewer: I would move this section above to the Out of the box experience and explain here how to program the ToF same as we are doing with motors and buttons.

Reviewer: yeap

**Read distance**

The `get_distance` function returns the distance readouts from the left, center-left, center, center-right, and right sensors.

**Outputs:**
`L`: Left sensor readout.
`CL`: Center-left sensor readout.
`C`: Center sensor readout.
`CR`: Center-right sensor readout.
`R`: Right sensor readout.

**Example Usage**

Here is an example of how to use the distance sensors to get distance measurements:

```python
from arduino_alvik import ArduinoAlvik
from time import sleep_ms
import sys

alvik = ArduinoAlvik()
alvik.begin()

while True:
    try:
        L, CL, C, CR, R = alvik.get_distance()
        T = alvik.get_distance_top()
        B = alvik.get_distance_bottom()
        print(f'T: {T} | B: {B} | L: {L} | CL: {CL} | C: {C} | CR: {CR} | R: {R}')
        sleep_ms(100)
    except KeyboardInterrupt as e:
        print('over')
        alvik.stop()
        sys.exit()
```

In this example, the robot uses its distance sensors to measure distances in its environment. The distance readouts from the left, center-left, center, center-right, right, top, and bottom sensors are printed continuously, allowing the robot to detect obstacles and measure distances accurately.

> ### Following a Line

Reviewer: I would move this section above to the Out of the box experience and explain here how to program the Line follower sensors same as we are doing with motors and buttons.

Reviewer: fine

### Grove Connectors

The Grove connectors on the Arduino Alvik robot are used for connectivity with compatible sensors and actuators following the Grove standard. The Grove system uses standardized connectors and pinouts to simplify the connection of various I2C devices. The system is modular and user-friendly, ideal for rapid prototyping and educational purposes. There are two Grove connectors, one on each side at the back of the robot, providing additional flexibility for expanding the robot's capabilities.

### Servomotor Connectors

The Arduino Alvik robot features 3-pin connectors for attaching servomotors while providing PWM control external servomotors. The connectors are disposed on a 6 by 2 header in the center at the back of the robot, supporting multiple servomotor connections. This arrangement enables complex robotic movements and functions, allowing the robot to perform a variety of additional tasks.

> 

Reviewer: Would [this image](https://docs.arduino.cc/static/766550ac1fea3d1bbc8f39d068a64b87/798d4/datasheet_connectors.png) help here?

Reviewer 2: Yes it is linked on 1472 but created individuals to the other sections

> 7. Test `bender.py`

Reviewer: let's get rid of the Bender example please XD, I am pretty sure we can make other nice stuff with the display instead of this one

Reviewer 2: Removed
